corehq/apps/style/static/style/js/bootstrap3/ui-element.js

```diff
@@ -0,0 +1,479 @@
+            };
+        }()),
+        map_list: (function() {
+            var KeyValList = function(guid, modal_title) {
+                var that = this;
+                eventize(this);
+                this.ui = $('<div class="enum-pairs" />');
+                this.value = {};
+                this.translated_value = {};
+                this.edit = true;
+                this.modal_id = 'enumModal-'+guid;
+                this.modal_title = modal_title;
+
+                this.$edit_view = $('<div class="well well-sm" />');
+                this.$noedit_view = $('<div />');
+                this.$formatted_view = $('<input type="hidden" />');
+                this.$modal_trigger = $('<a class="btn btn-default enum-edit" href="#'+this.modal_id+'" data-toggle="modal" />').html('<i class="fa fa-pencil"></i> Edit');
+
+                // Create new modal controller for this element
+                var $enumModal = $('<div id="'+this.modal_id+'" class="modal fade hq-enum-modal" />');
+                var $modalDialog = $('<div class="modal-dialog"/>');
+                var $modalContent = $('<div class="modal-content" />');
+
+                $modalContent.prepend('<div class="modal-header"><a class="close" data-dismiss="modal">×</a><h4 class="modal-title">Edit Mapping for '+this.modal_title+'</h4></div>');
+                var $modal_form = $('<form class="form-horizontal hq-enum-editor" action="" />'),
+                    $modal_body = $('<div class="modal-body" style="max-height:372px; overflow-y: scroll;" />');
+                $modal_body.append($('<fieldset />'));
+                $modal_body.append('<a href="#" class="btn btn-success" data-enum-action="add"><i class="fa fa-plus"></i> Add Key &rarr; Value Mapping</a>');
+
+                $modal_form.append($modal_body);
+                $modal_form.append('<div class="modal-footer"><button class="btn btn-primary" data-dismiss="modal">Done</button></div>');
+                $modalContent.append($modal_form);
+                $modalDialog.append($modalContent);
+                $enumModal.append($modalDialog);
+
+
+                $('#hq-modal-home').append($enumModal);
+
+                $('#'+this.modal_id).on('hide.bs.modal', function() {
+                    var $inputMap = $(this).find('form .hq-input-map'),
+                        pairs = {};
+                    for (var i=0; i < $inputMap.length; i++) {
+                        var key = $($inputMap[i]).find('.enum-key').val(),
+                            mapVal = $($inputMap[i]).find('.enum-value').val();
+                        if (key !== undefined){
+                            pairs[key] = mapVal.toString();
+                        }
+                    }
+                    that.val(pairs);
+                    that.fire('change');
+                });
+
+                $('#'+this.modal_id+' a').click(function() {
+                    if($(this).attr('data-enum-action') === 'add') {
```

Comment:
Nevermind about the removals, I see where those are handled. Although it'd be a tiny bit cleaner & create fewer click handlers to do removals in here, since this handler is still going to get called every time someone clicks remove.